homeassistant/components/lyric/config_flow.py

```diff
@@ -0,0 +1,142 @@
+class LyricFlowHandler(ConfigFlow):
+    """Handle a Lyric config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    def __init__(self):
+        """Initialize Lyric flow."""
+        self.client_id = None
+        self.client_secret = None
+        self.code = None
+        pass
+
+    async def async_step_user(self, code=None):
+        """Handle a flow initiated by the user."""
+        if self._async_current_entries():
+            return self.async_abort(reason='single_instance_allowed')
+
+        if not self.hass.data.get(DATA_LYRIC_CONFIG):
+            return self.async_abort(reason='no_config')
+
+        conf = self.hass.data.get(DATA_LYRIC_CONFIG)
+
+        self.client_id = conf[CONF_CLIENT_ID]
+        self.client_secret = conf[CONF_CLIENT_SECRET]
+
+        return await self.async_step_auth(code)
+
+    async def async_step_auth(self, code):
+        """Create an entry for auth."""
+        # Flow has been triggered from Lyric api
+        if isinstance(code, str):
+            return await self.async_step_code(code)
+
+        try:
+            with async_timeout.timeout(10):
+                url = await self._get_authorization_url()
+        except asyncio.TimeoutError:
+            return self.async_abort(reason='authorize_url_timeout')
+
+        return self.async_external_step(
+            step_id='auth',
+            url=url
+        )
+
+    async def _get_authorization_url(self):
+        """Get Lyric authorization url."""
+        client_id = self.client_id
+        client_secret = self.client_secret
+        redirect_uri = '{}{}'.format(
+            self.hass.config.api.base_url, AUTH_CALLBACK_PATH)
+        token_cache_file = self.hass.config.path(CONF_LYRIC_CONFIG_FILE)
+
+        lyric = Lyric(app_name='Home Assistant', client_id=client_id,
+                      client_secret=client_secret, redirect_uri=redirect_uri,
+                      token_cache_file=token_cache_file)
+
+        self.hass.http.register_view(LyricAuthCallbackView())
+
+        url = lyric.getauthorize_url
+
+        return url[:url.find('&state=') + 7] + self.flow_id
+
+    async def async_step_code(self, code):
+        """Received code for authentication."""
+        self.code = code
+        return self.async_external_step_done(next_step_id="creation")
+
+    async def async_step_creation(self, user_input=None):
+        """Create Lyric api and entries."""
+        client_id = self.client_id
+        client_secret = self.client_secret
+        redirect_uri = '{}{}'.format(
+            self.hass.config.api.base_url, AUTH_CALLBACK_PATH)
+        token_cache_file = self.hass.config.path(CONF_LYRIC_CONFIG_FILE)
+
+        lyric = Lyric(app_name='Home Assistant', client_id=client_id,
+                      client_secret=client_secret, redirect_uri=redirect_uri,
+                      token_cache_file=token_cache_file)
+
+        # pylint: disable=pointless-statement
+        lyric.getauthorize_url
```

Comment:
What happens here?

Reply:
I think this actually makes sure the lib sets up the api. The integration fails to work without it so I left it in. Not 100% sure why it is needed.

See: https://github.com/bramkragten/python-lyric/blob/master/lyric/__init__.py#L678

Reply:
The private api instance attribute isn't initialized until accessing this property, which could be considered a bug.

It would good if the library would be improved to have a more natural setup flow, to avoid having to disable pylint warnings. Maybe ask about this in a library repo issue?

Reply:
You create an instance of Lyric twice, once in `_get_authorization_url` and then again in `async_step_creation` might want to store the instance of `_get_authorization_url` and you don't have to do this.

The upstream code is old and needs improvement so PR's are welcome :-)

Reply:
Thanks @bramkragten. Should look better now :+1: